Task/FP-1375: Admin controlled messages

duckonomy opened the pull request:
## Overview: ##
Allow arbitrary and dismissable `SectionMessage` generation by admin (through `core/admin`) for components.

## Related Jira tickets: ##

* [FP-1375](https://jira.tacc.utexas.edu/browse/FP-1375)

## Summary of Changes: ##
1. Created a `/intromessages/custom/` endpoint (`GET`: Get templates & messages, `PUT`: Update `unread` state).
2. Create a `CustomMessage` UI component as a wrapper around `SectionMessages` to utilize `state.customMessages`.
3. Implemented necessary sagas/reducers to interact with the `/intromessages/custom/` endpoint.
4. Made messages of type `intro` have a dismiss button.

## Testing Steps: ##
0. Make sure to run migrations within the `core_portal_django` container: `./manage.py migrate`
1. [Make yourself staff or superuser](https://github.com/TACC/Core-Portal/wiki/How-to-Set-Your-User-as-Staff-or-Superuser).
2. Go to `core/admin`.
3. Create a *Custom message template*.
5. Go to the component you made it for (Will need refresh).
6. Ensure that it is correctly rendered
7. Repeat 3-5 for various different configurations (even multiple messages for single component) and ensure that they render accordingly
8. Ensure that dismissed (click `x`) messages do not show up.
9. Go to `core/admin` again.
10. Try removing some that aren't dismissed.
11. Refresh portal.
12. Ensure that the deleted messages do not show.

## UI Photos:
https://user-images.githubusercontent.com/9425579/159995718-6c0688db-797b-4c0a-9105-18c10f1fdca9.mov

## Notes: ##
1. (**For CMD**) Should there be a text limit for messages?
2. I wanted to refactor `intromessages` related components (both frontend/backend) into a more generic `messages` component. But wasn't sure how the migration process would go. Would there be a way to achieve this?
3. (**For CMD**) The `info` message type style does not seem to have an icon (`SectionMessage`, `Message`).  Thus, the dismiss button (`x`), which is a child of the icon does not render.

edmondsgarrett reviewed (approved):
LGTM, works as described.

rstijerina commented on server/portal/apps/intromessages/views.py:
```diff
@@ -34,3 +46,24 @@ def put(self, request, *args):
         return JsonResponse({'status': 'OK'})
+
+
+@method_decorator(login_required, name='dispatch')
+class CustomMessagesView(BaseApiView):
```
Similarly as above, would there be a case where we would want to access admin messages without authentication?

rstijerina commented on server/portal/apps/intromessages/models.py:
```diff
@@ -28,3 +28,59 @@ class IntroMessages(models.Model):
+    component = models.CharField(help_text='Component type', max_length=20, choices=COMPONENTS, default='Dashboard')
+    message_type = models.CharField(help_text='Message type', max_length=20, choices=MESSAGE_TYPES, default='info')
+    dismissible = models.BooleanField(default=True)
+    message = models.TextField(help_text='Message content', default='', blank=True)
```
The `TextField` supports newlines, but the custom message in the frontend does not render those newlines. Do we want to allow newlines for these messages?

rstijerina commented on client/src/components/_common/CustomMessage/CustomMessage.jsx:
```diff
@@ -0,0 +1,66 @@
+        <SectionMessage
+          type={template.message_type}
+          canDismiss={template.dismissible}
+          onDismiss={() => onDismiss(message)}
```
Not a part of this PR, but I noticed the close button is not vertically centered
<img width="105" alt="Screen Shot 2022-03-31 at 11 18 21 PM" src="https://user-images.githubusercontent.com/20326896/161194360-c43eb1f0-dae2-45ed-9584-e15257fdd127.png">

wesleyboar replied:
That problem will likely go away after [FP-546](https://jira.tacc.utexas.edu/browse/FP-546).

<sub>Via Inspector, I solved this by adding `display: flex;` to the button. And the new Button component has `display flex` on it. So, yay!</sub>

rstijerina reviewed (commented):
Nice work! Left a few comments, but overall 👍

rstijerina commented:
> I wanted to refactor intromessages related components (both frontend/backend) into a more generic messages component. But wasn't sure how the migration process would go. Would there be a way to achieve this?

@duckonomy Django supports [custom migrations](https://docs.djangoproject.com/en/4.0/howto/writing-migrations/), so one way to achieve that could be to deprecate the old model and migrate data over in a single new migration file. We've done something [similar in designsafe](https://github.com/DesignSafe-CI/portal/blob/main/designsafe/apps/googledrive_integration/migrations/0002_auto_20180215_2239.py#L13), though that was a bit simpler

rstijerina reviewed (changes requested):
Marking as "Request Changes" to keep track

wesleyboar commented on client/src/components/_common/Message/Message.module.scss:
```diff
@@ -61,7 +61,7 @@ p.is-scope-section {
 
 /* Design decided icon is not necessary for informational messages */
-.is-info .icon {
+.is-info .icon[aria-label='Notice'] {
```
Does the job. And I love attribute selectors.

But aria-label relies on text from markup which, if changed[^1], would be done without second thought to CSS, and thus this fix would be ineffectual. Less likely to change without someone checking for side-effects is a classname.

```suggestion
.is-info .icon:not(.close-icon) {
```

Untested.

[^1]: I admit the change would not happen under our nose (like translation), and is unlikely to happen (we rarely really think much about wording of accessible text), but it's good practice to avoid fragility.

duckonomy commented:
@rstijerina 
1. Refactored `/intromessages/` in the backend to a more generic `/messages/` folder. Also changed frontend `intro.sagas.js` and `intro.reducers.js` into `message.sagas.js` and `message.reducers.js`. 

I'm not sure how this will affect the existing migrations.. because there is a `dependency` on the moved file/directory.
```python
    dependencies = [
        ('message', '0001_initial'),
    ]
```

2. Adjusted `TextField` to `CharField` due to [your comment](https://github.com/TACC/Core-Portal/pull/615#discussion_r840216643) about allowing new lines and added 200 char limit in the according of our [conversation in slack](https://tacc-team.slack.com/archives/CQUA4D5KJ/p1648750276708879?thread_ts=1648155852.657219&cid=CQUA4D5KJ).

3. I wasn't sure how to allow resolve [this case](https://github.com/TACC/Core-Portal/pull/615#discussion_r840210198) at the moment. Currently, both `IntroMessages` and `CustomMessages` requires a relation to a user to keep track track of `unread` or dismissal status. So, routes that do not require authorization like `public-data` aren't supported for now. I'm not sure if we should allow these message would be allowed to be dismissed or purely admin-handled. If dismissable, do we do what we used to do and keep it in `localStorage`?. If this is required, I could make a ticket tackling this issue.

rstijerina commented:
> I wasn't sure how to allow resolve https://github.com/TACC/Core-Portal/pull/615#discussion_r840210198 at the moment. Currently, both IntroMessages and CustomMessages requires a relation to a user to keep track track of unread or dismissal status. So, routes that do not require authorization like public-data aren't supported for now. I'm not sure if we should allow these message would be allowed to be dismissed or purely admin-handled. If dismissable, do we do what we used to do and keep it in localStorage?. If this is required, I could make a ticket tackling this issue.

Yes, just what I was thinking too. We could use `localStorage` if the user is unauthenticated. New ticket sounds perfect 👍

duckonomy commented:
> > 5\. I wasn't sure how to allow resolve [Task/FP-1375: Admin controlled messages #615 (comment)](https://github.com/TACC/Core-Portal/pull/615#discussion_r840210198) at the moment. Currently, both IntroMessages and CustomMessages requires a relation to a user to keep track track of unread or dismissal status. So, routes that do not require authorization like public-data aren't supported for now. I'm not sure if we should allow these message would be allowed to be dismissed or purely admin-handled. If dismissable, do we do what we used to do and keep it in localStorage?. If this is required, I could make a ticket tackling this issue.
> 
> Yes, just what I was thinking too. We could use `localStorage` if the user is unauthenticated. New ticket sounds perfect 👍
👍 Created ticket [FP-1644](https://jira.tacc.utexas.edu/browse/FP-1644)

rstijerina commented on server/portal/apps/message/models.py:
```diff
@@ -0,0 +1,85 @@
+
+    component = models.CharField(help_text='Component type', max_length=20, choices=COMPONENTS, default='Dashboard')
+    message_type = models.CharField(help_text='Message type', max_length=20, choices=MESSAGE_TYPES, default='info')
+    dismissible = models.BooleanField(default=True)
```
> dismissible = models.BooleanField(default=True)

I wonder if `False` would be a better default - in the case of an admin adding a message and forgetting to mark it as non-dismissible, a user could dismiss the message before the admin fixed it as non-dismissible. What do you think?

rstijerina commented on server/portal/apps/message/models.py:
```diff
@@ -0,0 +1,85 @@
+        }
+
+    def __str__(self):
+        return "%s-%s-%s-%s" % (self.message_type, self.component,
```
Can we make this a bit more human readable? Currently it looks a bit wordy:
<img width="592" alt="Screen Shot 2022-05-11 at 3 04 27 PM" src="https://user-images.githubusercontent.com/20326896/167936956-df6eae12-9278-4d67-8632-234b7129c6c3.png">

duckonomy replied:
@rstijerina Would it be ok to just display details and not the actual message? I was worried that admin would have difficulty differentiating messages.

rstijerina reviewed (changes requested):
Looking great! Just a few more comments

duckonomy commented on server/portal/apps/message/views.py:
```diff
@@ -1,12 +1,13 @@
 """
-.. :module: apps.intromessages.views
+.. :module: apps.message.views
```
@rstijerina I've tried making it `messages` initially. But apparently there is a conflict with the built-in django `messages` utility if I make the app name that. So, it was throwing an error. So I opted to use `message`. I do think it would make more sense if the app name was plural. But I couldn't think of a generic name that could replace it.

rstijerina replied:
Ahh. What about announcements or portal_messages?

rstijerina replied:
Yeah! let's keep parity with the app

duckonomy commented:
@rstijerina 
1. Changed `introMessageName` to `messageName`
2. Changed batch behavior to single message update
3. Re-did migration
4. Changed route name to `portal_messages`
5. Improved readability of messages in admin panel.

rstijerina commented on client/src/components/Allocations/AllocationsLayout.jsx:
```diff
@@ -68,7 +68,7 @@ export const Layout = ({ page }) => {
     <Section
       bodyClassName="has-loaded-allocations"
-      introMessageName="ALLOCATIONS"
+      messageName="ALLOCATIONS"
```
This propagates down to the `IntroMessage` component, so this should probably stay as `introMessageName`

duckonomy replied:
@rstijerina Would this be true for `SectionMessages` too?
I wasn't sure what you meant by refactoring references to `introMessageName` in https://github.com/TACC/Core-Portal/pull/615#discussion_r870674573.

rstijerina replied:
@duckonomy Ah, sorry, that wasn't clear. I meant to say change references to `introMessageName` if that didn't pertain to the component it was in

rstijerina commented on client/src/components/_common/CustomMessage/CustomMessage.jsx:
```diff
@@ -0,0 +1,61 @@
+
+  function onDismiss(dismissMessage) {
+    dismissMessage.unread = false;
+    const payload = {
+      message: dismissMessage,
+    };
```
I noticed this sends the entire message object back over - think we can limit this to just the template id and unread status?

rstijerina reviewed (changes requested):
@duckonomy Looking great, just a couple more comments!

duckonomy commented:
@rstijerina 
1. Simplified payload (`templateId` and `unread`) for `PUT` request for updating custom messages.
2. Renamed `introMessages` to `introMessageComponents` and `introMessageName` to `messageComponentName` to reflect what it `introMessages` actually is (map of the state of a message for certain component). I think this makes more sense and is coherent with the use of the the prop in `Section` and `SectionMessages` for `customMessages` as well.

rstijerina commented on client/src/components/_common/Section/SectionMessages.jsx:
```diff
@@ -48,20 +48,28 @@ import './SectionMessages.css';
     /* FAQ: Alternate message name allows tracking custom message dismissal */
-    <IntroMessage messageName={introMessageName || introMessageText} canDismiss>
+    <IntroMessage
+      messageComponentName={messageComponentName || introMessageText}
+      canDismiss
+    >
       {introMessageContent}
     </IntroMessage>
   );
   const hasMessage =
-    isKnownIntroMessage(introMessageName) || children.length > 0;
+    isKnownIntroMessage(messageComponentName) || children.length > 0;
   const hasMessageClass = 'has-message';
 
+  const customMessage = (
+    <CustomMessage messageComponentName={messageComponentName}></CustomMessage>
+  );
+
```
@duckonomy I'm a little confused here. It seems both the `IntroMessage` and `CustomMessage` components are relying on the same `messageComponentName` parameter - is this preventing both an intro message and a custom message? My initial feeling is it would be clearer for `IntroMessage` to ingest `introMessageName`, and `CustomMessage` to ingest `customMessageName`

duckonomy replied:
@rstijerina So a separate prop?

rstijerina replied:
That's what I was thinking. Unless you think I'm missing something?

rstijerina commented:
> 2. Renamed `introMessages` to `introMessageComponents` and `introMessageName` to `messageComponentName` to reflect what `state.introMessages` actually is (map of the state of a message for certain component). I think this makes more sense and is coherent with the use of the the prop in `Section` and `SectionMessages` for `customMessages` as well.

@duckonomy
